src/threaddeadlock.h

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2019 Greg Griffith
+// Copyright (c) 2019 The Bitcoin Unlimited developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_THREAD_DEADLOCK_H
+#define BITCOIN_THREAD_DEADLOCK_H
+
+#include <boost/lexical_cast.hpp>
+#include <boost/thread.hpp>
+#include <inttypes.h>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <set>
+#include <string>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "utilstrencodings.h"
+
+enum LockType
+{
+    RECURSIVE, // CCriticalSection
+    SHARED, // CSharedCriticalSection
+    RECRUSIVESHARED, // CRecursiveSharedCriticalSection
+};
+
+#ifdef DEBUG_LOCKORDER // this ifdef covers the rest of the file
+#include <sys/syscall.h>
+#ifdef __linux__
+inline uint64_t getTid(void)
```

Comment:
Suggest renaming `getTid` to `getThreadID` and perhaps move to `utilprocess`

Reply:
The function name is intentional to match the linux system call (used in the linux define)

Reply:
Hey I have an extant trivial PR that fixes an issue with this code #1905 that I encountered when compiling for Darwin.

My notes from that PR:

Shouldn't the `<sys/syscall.h>` be **inside** the `__linux__`  `ifdef`?  Or.. am I missing something? By my reckoning not all platforms have that header.... or do they? (I honestly don't know).  Sorry.. perhaps I'm out of it here, and they do.  But on Windows MinGW (which is basically the platform used to build Windows) that header is nowhere to be found....?

Also, below, in the non-Linux case: 

- I had bad_lexical_cast exceptions raised on Darwin here (that platform doesn't have what one expects as a thread id, basically).... I ended up going with something that uses `std::hash` to transform the opaque `std::thread::id` type into an integer (which is guaranteed to work by the standard)... This does produces ugly log messages if you were to log the value, but at least it's unique-ish and useful for placing in eg a map.